[RFC][LangRef] Specify that the accessed bytes of concurrent atomics must be either disjoint or the same (#204329)
llvm/llvm-project: llvm/docs/LangRef.md

-  Otherwise, if R{sub}`byte` may see exactly one write,
   R{sub}`byte` returns the value written by that write.
-  Otherwise, if R is atomic, and all the writes R{sub}`byte` may
   see are atomic, and R and the writes all access the exact same set of
   bytes, it chooses one of the values written. See the {ref}`Atomic
   Memory Ordering Constraints <ordering>` section for additional
   constraints on how the choice is made. Targets may impose additional
   requirements on R and the writes it may see based on their `syncscope`.
-  Otherwise R{sub}`byte` returns `undef`.

> Defined atomic accesses cannot tear: If a byte subaccess R{sub}`byte1`

Reviewer: What does "defined atomic accesses" mean?

Author: Those that aren't involved in a data race such that the read returns undef.
The "Defined atomic accesses cannot tear:" part really is only informational, and the point of "Defined" in there is to not appear to promise something about atomics that race with non-atomics or are racing because of mismatching syncscopes. Those can still appear as if they are tearing in an execution, as is allowed for undef values.

The most rigorous solution to this vagueness would probably be to remove the "Defined atomic accesses cannot tear:" part and let the constraint speak for itself, but that might make it more difficult to follow. Do you have a preference?

Reviewer 2: You could just remove the first word?
```suggestion
Atomic accesses cannot tear: If a byte subaccess R{sub}`byte1`
```
A read that returns undef/poison doesn't really "tear", it doesn't return any data at all.

Reviewer: I like Ralf's suggestion.

Author: Done.

-  Otherwise, if R{sub}`byte` may see exactly one write,
   R{sub}`byte` returns the value written by that write.
-  Otherwise, if R is atomic, and all the writes R{sub}`byte` may
   see are atomic, and R and the writes all access the exact same set of
   bytes, it chooses one of the values written. See the {ref}`Atomic
   Memory Ordering Constraints <ordering>` section for additional
   constraints on how the choice is made. Targets may impose additional
   requirements on R and the writes it may see based on their `syncscope`.
-  Otherwise R{sub}`byte` returns `undef`.

Defined atomic accesses cannot tear: Two byte subaccesses R{sub}`byte1`,
R{sub}`byte2` of an atomic read R cannot read from different atomic

> writes W{sub}`1`, W{sub}`2` if both read subaccesses, R{sub}`byte1`,

Reviewer 2: ```suggestion
writes W{sub}`1`, W{sub}`2` even if both read subaccesses, R{sub}`byte1`,
```
Seems more clear to me this way?

Author: No, that breaks the intended meaning: if Rbyte1 and Rbyte2 may not both see both writes W1 and W2, the read subaccesses don't have to read from the same write. For example:
```
S1: store atomic to bytes {0, 1, 2, 3} starting at address p
S2: store atomic to bytes {0, 1} starting at address p
L: load atomic from bytes {0, 1, 2, 3} starting at address p
```
These three operations are in a single thread, so they are in happens-before. `L` should be allowed to read bytes 0, 1 from `S2` and bytes 2, 3 from `S1`.

Reviewer 2: We already say that you cannot read from a write that is superseded by a later write that you can see, so for that example I don't see why any clarification would be needed. There's only one possible write to read from for each read subaccess.

Author: If we say, like you suggested:

> Two byte subaccesses R{sub}`byte1`, R{sub}`byte2` of an atomic read R cannot
> read from different atomic writes W{sub}`1`, W{sub}`2` **even** if both read subaccesses,
> R{sub}`byte1`, R{sub}`byte2`, may see both writes, W{sub}`1`, W{sub}`2`.

Then, that now implies this statement (at least to me):

> Two byte subaccesses R{sub}`byte1`, R{sub}`byte2` of an atomic read R cannot
> read from different atomic writes W{sub}`1`, W{sub}`2`.

But that means for my example above that the subaccesses of L for byte 0 and byte 3 are not allowed to read from different writes. `S2` doesn't provide byte 3, so they'd both need to read from `S1`, but that would violate the requirement that you can't read from superseded writes.
So, the constraints would be unsatisfiable for this case.

Reviewer 2: I think what I am missing is which problem this paragraph is trying to solve. If I had to state a no-tear guarantee that only applies for the case where there are actual races, I'd say something like:

All byte subaccesses of an atomic read R must read from the same atomic write, except for the case where all byte subaccesses are reading from writes that happen-before R.

Author: > All byte subaccesses of an atomic read R must read from the same atomic write, except for the case where all byte subaccesses are reading from writes that happen-before R.

That breaks for the case where you have two writes W1 and W2 that both happen before R (but there is no hb edge in either direction between the two writes). In that case, all subaccesses of R should read from the same Write. The monotonic modification order is not enough to rule that out since LLVM has `unordered` atomic accesses that are not ordered by the modification order.

Reviewer 2: > That breaks for the case where you have two writes W1 and W2 that both happen before R (but there is no hb edge in either direction between the two writes). 

And I assume those writes are partially overlapping? In that case we have UB. No read has to even happen to cause UB here.

Author: > > That breaks for the case where you have two writes W1 and W2 that both happen before R (but there is no hb edge in either direction between the two writes).
> 
> And I assume those writes are partially overlapping? In that case we have UB. No read has to even happen to cause UB here.

No, sorry, I should have clarified: W1, W2 and R all access the exact same set of bytes. So no data race.

Reviewer 2: Oh, of course.

I find it kind of confusing how this defines first the per-byte accesses and then needs this no-tearing clause when combining them to whole accesses. But that's how the entire docs are already written so 🤷 .

I am a bit concerned about the case where byte 0 can see W1 and W2, and byte 1 can see W1 and W3. It seems like your definition would allow the sub-reads to use any combination of different writes. But maybe that situation cannot arise because it would have been a write-write race somewhere before already?

I think under the UB you have, your definition is equivalent to the following that I find easier to understand -- or did I miss something?
> Defined atomic accesses cannot tear: If a byte subaccess R{sub}`byte1` of
an atomic read R reads from W{sub}`1`, then all other byte subaccesses R{sub}`byte2` of R that can see  W{sub}`1` must also read from W{sub}`1`.

Author: > I am a bit concerned about the case where byte 0 can see W1 and W2, and byte 1 can see W1 and W3. It seems like your definition would allow the sub-reads to use any combination of different writes. But maybe that situation cannot arise because it would have been a write-write race somewhere before already?

Yes, that situation can't happen: If byte 0 can see W1 and W2 and byte 1 can see W1 and W3, then W1 overlaps partially with W2 and with W3 and they are not in hb. The other addition of this PR makes involved reads therefore return undef.
(I suppose there are also other situations where you can get the may-see situation you describe with additional partially overlapping writes that overshadow W2 and W3 partially, I believe that would also cause undef reads).
If we wanted to define partially overlapping atomics to behave properly atomically (which we currently don't), we would need additional constraints.

>I think under the UB you have, your definition is equivalent to the following that I find easier to understand -- or did I miss something?
>
>>Defined atomic accesses cannot tear: If a byte subaccess R{sub}byte1 of
an atomic read R reads from W{sub}1, then all other byte subaccesses R{sub}byte2 of R that can see W{sub}1 must also read from W{sub}1.

Yes, I think that works, and I like the simpler formulation. I've updated the PR (with minor adjustments: W{sub}1 doesn't need an index and it should be an atomic write).

For the record: There would be a problem with it if partially overlapping atomics were supposed to behave properly atomically (again: which we currently don't want):
Then, this constraint would (in my opinion: wrongly) disallow in the following scenario that L reads anything from S2:

```
thread 1:
  S1: store atomic unordered to bytes {0, 1, 2, 3} starting at %p
  Rel: store atomic release %flag

thread 2:
  S2: store atomic unordered to bytes {0, 1} starting at %p
  Acq: load atomic acquire %flag // assume it reads from Rel and establishes synchronization and hb
  L: load atomic unordered from bytes {0, 1, 2, 3} starting at %p  // should be allowed to read {0, 1} from S2 and {2, 3} from S1 if partially overlapping atomics behaved atomically
```

Reviewer 2: > There would be a problem with it if partially overlapping atomics were supposed to behave properly atomically

Yes, I was working under the same assumption as you, i.e. that this is not supported at the moment.

For a simpler introduction to the ordering constraints, see the
{doc}`Atomics`.

For the following, we call two or more accesses *perfectly overlapping*
if they all access the exact same set of bytes, i.e., they access the
same address and have the same access size. By the constraints of the

> previous section, racing atomic accesses must be perfectly overlapping

Reviewer 2: I suppose "racing" here is meant to exclude the case where both accesses are reads?

"racing" could refer to "race condition" or "data race", which in my experience typically have different definitions:
- data race: "two overlapping accesses that are not happens-before ordered where at least one is a write"
- race condition: "a data race where at least one of the involved accesses is non-atomic"

So "data race" would have the right meaning here. But I don't know if LLVM has its own definition of these terms.

Reviewer 2: Maybe it doesn't matter too much here, since this is just a recap anyway. "act atomically" is also not defined, and if they are all reads than "acting non-atomically" is actually entirely fine.

But I think I'd find this a bit less confusing:
```suggestion
previous section, overlapping atomic accesses that are not ordered by happens-before must be perfectly overlapping
```

Author: Sounds good to me, updated. It's better to not use the word "racing" here. Thanks!

> "act atomically" is also not defined, and if they are all reads than "acting non-atomically" is actually entirely fine.

Yes, I'd say that we don't need to call the situation with reads out separately here because of that. I wouldn't want to go into more detail/formalism than "act atomically" here because then we'd just repeat the previous section :)
